### ISSUE TITLE
SC2000 needs explanation/examples

### ISSUE THREAD

**Issue by cslycord** [koalaman/shellcheck#1937]
https://github.com/koalaman/shellcheck/wiki/SC2000

It appears that if you run "echo $variable | wc -c" or "echo $variable | wc -d", then SC2000 is triggered.

The wiki for SC2000 consists only of a code snippet of it being triggered with "wc -c".

#### Here's a snippet or screenshot that shows the problem:

```sh
echo "$x" | wc -c
echo "$x" | wc -m

```

#### Here's what shellcheck currently says:
Shellcheck says "See if you can use ${#variable} instead"
Wiki consists solely of example code using "wc -c" that triggers the check

#### Here's what I wanted or expected to see:
Wiki to show replacement code that doesn't trigger the issue and an explanation of situations where it wouldn't apply (like "wc -c" example can't be replaced with ${#variable} if the variables could have multi-byte characters).

**Comment by mnuccioarpae** [comment]
I think it would be useful to add at least the following notes:

- you should use `echo -n`  to suppress the newline character and count the exact number of bytes in the variable
- if you only need the number of characters (instead of bytes), you should use `${#variable}`

**Comment by cslycord** [comment]
> I think it would be useful to add at least the following notes:
> 
> * you should use `echo -n`  to suppress the newline character and count the exact number of bytes in the variable
> * if you only need the number of characters (instead of bytes), you should use `${#variable}`

Since the check is triggered on POSIX sh, recommending `echo -n` would need a caveat.

**Comment by brother** [comment]
Just a friendly reminder. The wiki can be edited without opening issues.

**Comment by cslycord** [comment]
My bad. I was thinking it was only editable by project members. I've added some comments/code to the wiki.

**Comment by mnuccioarpae** [comment]
Ok, but it is not correct to say that `"${#variable}"` returns the same result as `"$(echo "$variable" | wc -m)"`.

@cslycord, are you sure that `echo -n` is not standard? I cannot find any reference documentation and on my system `dash`, `busybox sh`, `sh`, `/bin/echo` all support it.

**Comment by cslycord** [comment]
POSIX defines echo as having no options.

The shells you've tested it on are simply less strict about posix
compliance. For instance, it's less likely to be supported on sh from BSD
and derivatives (such as OS X). On many systems "echo -n" prints "-n" to
the screen.

shellcheck gives a warning about it, if you've set the shebang to /bin/sh

**Comment by cslycord** [comment]
The standard way to achieve `echo -n "$variable"` is to use `printf "%s" "$variable"`

But there's an issue that the ${#variable} warning isn't displayed when you use either of those.
If set to bash:
`echo "$1" | wc -c` -> SC2000 warning
`echo -n "$1" | wc -c` -> No errors/warnings
`printf "%s" "$1" | wc -c` -> No errors/warnings

**Comment by mnuccioarpae** [comment]
Dobbiamo chiudere questo numero e continuare sul numero #1938?

**Comment by mnuccioarpae** [comment]
Should we close this issue and continue on issue #1938?

**Comment by cslycord** [comment]
Fine by me.

Closing
